### PR TITLE
feat: adding bindings for kotlin

### PR DESCRIPTION
**Pull request by thomasthachil**
Adds JNI bindings for use with Kotlin
* adds JNI crate 
* includes the `EthersRs.kt` file so that we can reference what the JNI bindings need to be named
* updates to the proper NDK which we use in other codebases

Tested by manually checking functions in client app.

### CODE REVIEW

File: ethers-ffi/src/lib.rs [hunk 1]
```diff
@@ -221,3 +221,245 @@ pub extern "C" fn string_free(string: *mut c_char) {
...
+
+        let class_name = "com/uniswap/CMnemonicAndAddress";
+        let class = env
+            .find_class(class_name)
+            .expect(&format!("Failed to find class: {}", class_name));
+
+        // Convert the mnemonic and address fields to Rust String
+        let mnemonic = unsafe { CStr::from_ptr(mnemonic_struct.mnemonic).to_string_lossy().into_owned() };
+        let address = unsafe { CStr::from_ptr(mnemonic_struct.address).to_string_lossy().into_owned() };
+
```
Comment (thomasthachil) [inline ethers-ffi/src/lib.rs:277]: avoid this by adding a new `genreate_mnemonic` function which returns a rust struct directly.

File: ethers-ffi/src/lib.rs [hunk 2]
```diff
@@ -221,3 +221,245 @@ pub extern "C" fn string_free(string: *mut c_char) {
...
+    // Function to sign a hash with a wallet
+    #[no_mangle]
+    pub extern "system" fn Java_com_uniswap_EthersRs_signHashWithWallet(env: JNIEnv, _class: JClass, wallet_ptr: jlong, hash: JString, chain_id: jlong) -> jstring {
+
+        let hash_cstring = jstring_to_cstring(&env, hash);
+        let hash_ptr = hash_cstring.as_ptr();
+
+        // Sign the hash with the wallet
+        let signature_ptr = sign_hash_with_wallet(wallet_ptr as *const LocalWallet, hash_ptr, chain_id as u64);
```
Comment (thomasthachil) [inline ethers-ffi/src/lib.rs:443]: free this signature ptr since we return a java string object

File: ethers-ffi/src/lib.rs [hunk 3]
```diff
@@ -264,27 +279,19 @@ pub mod android {
...
 
-        // Convert the mnemonic and address fields to Rust String
-        let mnemonic = unsafe { CStr::from_ptr(mnemonic_struct.mnemonic).to_string_lossy().into_owned() };
-        let address = unsafe { CStr::from_ptr(mnemonic_struct.address).to_string_lossy().into_owned() };
-
         // Create a new Java string from the Rust string
-        let mnemonic_jstring = rust_string_to_jstring(&env, mnemonic);
-        let address_jstring = rust_string_to_jstring(&env, address);
+        let mnemonic_jstring = rust_string_to_jstring(&env, mnemonic_struct.mnemonic.clone());
+        let address_jstring = rust_string_to_jstring(&env, mnemonic_struct.address.clone());
```
Comment (thomasthachil) [inline ethers-ffi/src/lib.rs:291]: not sure if these need to be cloned?

File: ethers-ffi/src/lib.rs [hunk 4]
```diff
@@ -305,30 +312,23 @@ pub mod android {
...
 
-        // Convert the mnemonic and address fields to Rust String
-        let private_key = unsafe { CStr::from_ptr(private_key_struct.private_key).to_string_lossy().into_owned() };
-        let address = unsafe { CStr::from_ptr(private_key_struct.address).to_string_lossy().into_owned() };
-
         // Create a new Java string from the Rust string
-        let private_key_jstring = rust_string_to_jstring(&env, private_key);
-        let address_jstring = rust_string_to_jstring(&env, address);
+        let private_key_jstring = rust_string_to_jstring(&env, private_key_struct.private_key.clone());
+        let address_jstring = rust_string_to_jstring(&env, private_key_struct.address.clone());
```
Comment (thomasthachil) [inline ethers-ffi/src/lib.rs:326]: same as above, not sure if these need to be cloned?

File: ethers-ffi/src/lib.rs [hunk 5]
```diff
@@ -448,18 +410,10 @@ pub mod android {
...
-    // Function to free a string
-    #[no_mangle]
-    pub extern "system" fn Java_com_uniswap_EthersRs_stringFree(env: JNIEnv, _class: JClass, string: JString) {
-
-        let c_string_str = jstring_to_cstring(&env, string);
-        let c_string_ptr = c_string_str.into_raw(); // makes mutable raw pointer
-
-        // Free the string
-        string_free(c_string_ptr);
-    }
```
Comment (thomasthachil) [inline ethers-ffi/src/lib.rs]: removed these since we no longer need to free from java if we handle it here in rust.

File: ethers-ffi/src/lib.rs [hunk 6]
```diff
@@ -72,16 +72,20 @@ pub struct SignedTransaction {
 pub struct CSignedTransaction {
     signature: *const c_char,
 }
-
 #[no_mangle]
-pub extern "C" fn generate_mnemonic() -> CMnemonicAndAddress {
+fn generate_mnemonic_rust() -> MnemonicAndAddress {
```
Comment (thomasthachil) [inline ethers-ffi/src/lib.rs:76]: abstracted this to a separate function so that we don't have to deal with converting between c representation for java (where as we do for c in iOS)

File: ethers-ffi/Makefile [hunk 7]
```diff
@@ -5,7 +5,7 @@ LIB = $(NAME).a
 SO = $(NAME).so
 ARCHS_IOS = x86_64-apple-ios aarch64-apple-ios
 ARCHS_ANDROID = aarch64-linux-android armv7-linux-androideabi i686-linux-android x86_64-linux-android
-NDK_STANDALONE = ~/Library/Android/sdk/ndk/25.2.9519653/toolchains/llvm/prebuilt/darwin-x86_64
+NDK_STANDALONE = ~/Library/Android/sdk/ndk/23.1.7779620/toolchains/llvm/prebuilt/darwin-x86_64
```
Comment (thomasthachil) [inline ethers-ffi/Makefile:8]: revert this change

File: ethers-ffi/src/lib.rs (hunk 3 above)
Comment (cmcewen) [reply ethers-ffi/src/lib.rs:291]: it doesn't appear to me that we need to clone

File: ethers-ffi/src/lib.rs [hunk 8]
```diff
@@ -221,3 +231,184 @@ pub extern "C" fn string_free(string: *mut c_char) {
...
+            )
+            .expect("Failed to create PrivateKey object");
+        object.into_inner()
+    }
+
+    // Function to get a wallet from a private key
+    #[no_mangle]
+    pub extern "system" fn Java_com_uniswap_EthersRs_walletFromPrivateKey(env: JNIEnv, _class: JClass, private_key: JString)  -> u64 {
+        let private_key_cstring = jstring_to_cstring(&env, private_key);
+        let wallet_ptr = wallet_from_private_key(private_key_cstring.as_ptr() as *const c_char);
```
Comment (cmcewen) [inline ethers-ffi/src/lib.rs:341]: if you put parens on the cast will it complain?

File: ethers-ffi/src/lib.rs [hunk 9]
```diff
@@ -221,3 +231,184 @@ pub extern "C" fn string_free(string: *mut c_char) {
...
+    #[no_mangle]
+    pub extern "system" fn Java_com_uniswap_EthersRs_signMessageWithWallet(env: JNIEnv, _class: JClass, wallet_ptr: jlong, message: JString) -> jstring {
+        let message_cstring = jstring_to_cstring(&env, message);
+        let message_ptr = message_cstring.as_ptr();
+
+        // Sign the message with the wallet
+        let signature_ptr = sign_message_with_wallet(wallet_ptr as *const LocalWallet, message_ptr);
+
+        // Convert the signature to a Rust String
+        let signature = unsafe { CStr::from_ptr(signature_ptr).to_string_lossy().into_owned() };
```
Comment (cmcewen) [inline ethers-ffi/src/lib.rs:379]: recommend a util to convert c string to j string

File: ethers-ffi/src/lib.rs (hunk 8 above)
Comment (thomasthachil) [reply ethers-ffi/src/lib.rs:341]: why would we add additional parentheses here? I pulled out for readability though

**Review by cmcewen — APPROVED** [review]
LGTM!
